data/src/main/java/com/microsoft/azure/kusto/data/auth/CloudInfo.java

```diff
@@ -0,0 +1,132 @@
+
+import org.apache.http.HttpHeaders;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+class CloudInfo {
+    public static final String METADATA_ENDPOINT = "v1/rest/auth/metadata";
+    public static final String DEFAULT_KUSTO_CLIENT_APP_ID = "db662dc1-0cfe-4e1c-a843-19a68e65be58";
+    public static final boolean DEFAULT_LOGIN_MFA_REQUIRED = false;
+    public static final String DEFAULT_PUBLIC_LOGIN_URL = "https://login.microsoftonline.com";
+    public static final String DEFAULT_REDIRECT_URI = "https://microsoft/kustoclient";
+    public static final String DEFAULT_KUSTO_SERVICE_RESOURCE_ID = "https://kusto.kusto.windows.net";
+    public static final String DEFAULT_FIRST_PARTY_AUTHORITY_URL = "https://login.microsoftonline.com/f8cdef31-a31e-4b4a-93e4-5f571e91255a";
+    public static final CloudInfo DEFAULT_CLOUD = new CloudInfo(
+            DEFAULT_LOGIN_MFA_REQUIRED,
+            DEFAULT_PUBLIC_LOGIN_URL,
+            DEFAULT_KUSTO_CLIENT_APP_ID,
+            DEFAULT_REDIRECT_URI,
+            DEFAULT_KUSTO_SERVICE_RESOURCE_ID,
+            DEFAULT_FIRST_PARTY_AUTHORITY_URL
+    );
+
+    private static final Map<String, CloudInfo> cache = new HashMap<>();
+
+    private final boolean loginMfaRequired;
+    private final String loginEndpoint;
+    private final String kustoClientAppId;
+    private final String kustoClientRedirectUri;
+    private final String kustoServiceResourceId;
+    private final String firstPartyAuthorityUrl;
+
+    public CloudInfo(boolean loginMfaRequired, String loginEndpoint, String kustoClientAppId, String kustoClientRedirectUri, String kustoServiceResourceId, String firstPartyAuthorityUrl) {
+        this.loginMfaRequired = loginMfaRequired;
+        this.loginEndpoint = loginEndpoint;
+        this.kustoClientAppId = kustoClientAppId;
+        this.kustoClientRedirectUri = kustoClientRedirectUri;
+        this.kustoServiceResourceId = kustoServiceResourceId;
+        this.firstPartyAuthorityUrl = firstPartyAuthorityUrl;
+    }
+
+    public static void manuallyAddToCache(String clusterUrl, CloudInfo cloudInfo){
+        synchronized (cache) {
+            cache.put(clusterUrl, cloudInfo);
+        }
+    }
+
+    public static CloudInfo retrieveCloudInfoForCluster(String clusterUrl) throws DataServiceException {
+        synchronized (cache) {
+            CloudInfo cloudInfo = cache.get(clusterUrl);
+            if (cloudInfo != null) {
+                return cloudInfo;
+            }
+
+            CloudInfo result;
+
+            try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
+                HttpGet request = new HttpGet(clusterUrl + "/" + METADATA_ENDPOINT);
```

Comment:
What if clusterUrl ends with a '/'

Reply:
Thanks!

Technically you could say the same comments about when I concatenate to URIs I get from the metadata query,
but for them even c# relies on them not having a '/'. But maybe I should do it anyway for consistency? WDYT?

Reply:
Aren't there any Java URL composition APIs which do this without having to manually parse the last char?

Reply:
I really hate how github only displays the outdated diffs, you'll see in a recent commit that I changed it to 
`HttpGet request = new HttpGet(new URIBuilder(clusterUrl).setPath(METADATA_ENDPOINT).build().toString());`

Which is more verbose but is the "correct" way to do it.

The question is should I do the same for other URI concatinations we have, even if we know for sure they don't end with '/'?

Reply:
yea while you are on it

Reply:
LMK if the code looks good now

Reply:
fine with me